library/network_connections.py

```diff
@@ -993,16 +995,24 @@ def connection_create(self, connections, idx, connection_current=None):
                     NM.SETTING_IP_CONFIG_ROUTE_METRIC, ip["route_metric4"]
                 )
             for d in ip["dns"]:
-                if d["family"] == socket.AF_INET:
-                    s_ip4.add_dns(d["address"])
-            for s in ip["dns_search"]:
-                s_ip4.add_dns_search(s)
-            s_ip4.clear_dns_options(True)
-            for s in ip["dns_options"]:
-                s_ip4.add_dns_option(s)
+                if not ipv4_is_disabled:
+                    if d["family"] == socket.AF_INET:
+                        s_ip4.add_dns(d["address"])
+                        for s in ip["dns_search"]:
+                            s_ip4.add_dns_search(s)
+                        s_ip4.clear_dns_options(True)
+                        for s in ip["dns_options"]:
+                            s_ip4.add_dns_option(s)
+                else:
+                    if d["family"] == socket.AF_INET:
```

Comment:
the current version iterates over all `ip["dns"]`, and whenever it find a suitable IPv6 address, it does:
```
    for s in ip["dns_search"]:
        s_ip6.add_dns_search(s)
```
That's not really a problem, because `add_dns_search()` will not add duplicate entries. But it still seems odd. Can we do this outside of the loop?

How about to unwrap this loop to:

```python
    for d in ip["dns"]:
        if d["family"] != socket.AF_INET6:
            continue
        if ipv6_is_disabled:
            raise MyError(
                         "when IPv6 is disabled, DNS setting for IPv6 nameservers is not allowed."
            )
        s_ip6.add_dns(d["address"])
    if not ipv6_is_disabled:
       for s in ip["dns_search"]:
           s_ip6.add_dns_search(s)
        s_ip6.clear_dns_options(True)
        for s in ip["dns_options"]:
            s_ip6.add_dns_option(s)
```
(which has a different behavior, of also configuring the DNS option/searches, if there are not IPv6 name servers. But that's fine, you might get the nameserver via DHCP).

Reply:
@thom311 , do you mean that we can get nameserver no matter what ? Even if `ip["dns"] == []`, we can always get nameserver via DHCP ? So, as long as `ipv6.method != disabled`, we should configure the DNS option/searches.

Reply:
@thom311 , do you mean that we can get nameserver no matter what ? Even if `ip["dns"] == []`, we can always get nameserver via DHCP ? So, as long as `ipv6.method != disabled`, we should configure the DNS option/searches.

Reply:
yes, that's what I meant.